internal/controller/application_controller.go

```diff
@@ -163,49 +174,142 @@ func (r *ApplicationPullServiceAccountCreator) getImageRepositoryPullSecretNames
-		log.Error(err, "failed to create service account", "serviceAccountName", l.Action, l.ActionAdd)
-		return err
+	foundAggregatedSecret := &corev1.Secret{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: aggregatedSecretName, Namespace: application.Namespace}, foundAggregatedSecret)
+
+	if err != nil && errors.IsNotFound(err) {
+		log.Info("Creating aggregated pull secret.")
+		if err := r.Client.Create(ctx, aggregatedSecret); err != nil {
+			log.Error(err, "failed to create aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionAdd)
+			return nil, err
+		}
+		log.Info("Aggregated pull secret created successfully.")
+		return aggregatedSecret, nil
+	} else if err != nil {
+		log.Error(err, "failed to get aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionView)
+		return nil, err
+	}
+
+	// If secret already exists, check if update is needed
+	if string(foundAggregatedSecret.Data[corev1.DockerConfigJsonKey]) != string(marshaledData) {
+		log.Info("Aggregated pull secret content changed, updating.")
+		foundAggregatedSecret.Data = aggregatedSecret.Data // Update the data
+		// Ensure owner references are maintained if the secret already existed
+		if err := controllerutil.SetOwnerReference(application, foundAggregatedSecret, r.Scheme); err != nil {
+			log.Error(err, "failed to update owner reference for existing aggregated secret", "applicationName", application.Name)
+			return nil, err
+		}
+		if err := r.Client.Update(ctx, foundAggregatedSecret); err != nil {
+			log.Error(err, "failed to update aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionUpdate)
+			return nil, err
+		}
+		log.Info("Aggregated pull secret updated successfully.")
+	} else {
+		log.Info("Aggregated pull secret content is up-to-date.")
+	}
+
+	return foundAggregatedSecret, nil
+}
+
+// updateServiceAccountWithAggregatedSecret updates the ServiceAccount to include
+// the aggregated pull secret as an imagePullSecret.
+func (r *AggregatedPullSecretCreator) updateServiceAccountWithAggregatedSecret(ctx context.Context, sa *corev1.ServiceAccount, aggregatedSecret *corev1.Secret) error {
+	log := ctrllog.FromContext(ctx)
+
+	// Check if the aggregated pull secret is linked to the SA
+	found := false
+	for _, pullSecret := range sa.ImagePullSecrets {
+		if pullSecret.Name == aggregatedSecret.Name {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		log.Info("Aggregated pull secret not linked to Service Account, adding it.")
+		sa.ImagePullSecrets = append(sa.ImagePullSecrets, corev1.LocalObjectReference{Name: aggregatedSecret.Name})
+
+		if err := r.Client.Update(ctx, sa); err != nil {
+			log.Error(err, "failed to update Service Account with aggregated pull secret", l.Action, l.ActionUpdate)
+			return err
+		}
+		log.Info("Service Account updated successfully with aggregated pull secret.")
+	} else {
+		log.Info("Service Account already links the aggregated pull secret.")
```

Comment:
don't log this, because when controller restarts it will process again ALL applications, so this will just pollute logs